sessions/2018/september/Session_13.md

```diff
@@ -0,0 +1,138 @@
+# SOFTWARE ENGINEERING IN JAVA
+
+## Session 13 (25/09/2018)
+
+- Lists, Maps and Generics
+- Gradle
+- Spring Initializr
+
+### Notes:
+
+#### Lists
+
+When we define a List in Java we do in the following way:
+
+```java
+List<String> list = new ArrayList<>();
+list.add("asdf");
+list.add("123");
+list.add("qwer");
+```
+Here, we are creating an ArrayList (is the most common used implementation of Java `List`) of `Strings` and adding some values. There is a lot of methods that Java integrates to allow the manipulation of `Lists`.
+
+A quick way to loop over the elements, introduced in Java 8, is using forEach:
+```java
+list.forEach(System.out::println);
+```
+
+We also could to define list datatype as `Object`:
+```java
+int a=2;
+List<Object> list_b = Arrays.asList(new GenericTest(), "asdf", 123, a);
+```
+In this example, all that we are storing in the list is an object even the `String "asdf"`, the `Integer 123` and even the primitive `int a` variable, because from Java 7 this values are intelligently autoboxed to map towards the List datatype, so at the end is stored as an `Integer` object.
+
+#### Maps
+
+Maps are structures to keep data in the <key, value> form. The definition, usage and iteration over are similar to the way to handle `Lists` and also has a lot of built-in methods to manipulate and make different operations over our `Maps`:
+
+```java
+Map<String, Integer> map = new HashMap<>();
+map.put("a", 1);
+map.put("b", 2);
+map.forEach((k,v)-> System.out.println(k+v));
+```
+Something to notice in the previous `print` statement, is that two different datatapes are combined using the `+` operator, then both values will be treated as `String` since one of them is. So a string is printed.
```

Comment:
It allows you to concatenate because the *first* parameter is a String. I don't think it would work if the second parameter was a String, but not the first.

Reply:
Okay, thank you